### PR TITLE
MNT: Add asv benchmarking scripts.

### PR DESCRIPTION
**Pull request by danielballan**
To follow on a conversation with @klauer this morning, I configured the benchmarking utility ASV ("air speed velocity" -- it's a Monty Python reference). It runs simple functions in a clean environment, and it can track speed, memory usage, etc. over the commit history of a project. The output can be simple or [sophisticated](http://mdboom.github.io/astropy-benchmark/).

To start, I benchmarked the time it takes to insert 1000 Events.

```
$ asv run
· Fetching recent changes
· Creating environments
· Discovering benchmarks
·· Uninstalling from py3.4-humanize-mongoengine0.8.7-numpy1.9-prettytable-pytz-pyyaml-six
·· Building for py3.4-humanize-mongoengine0.8.7-numpy1.9-prettytable-pytz-pyyaml-six
·· Installing into py3.4-humanize-mongoengine0.8.7-numpy1.9-prettytable-pytz-pyyaml-six..
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] · For metadatastore commit hash 2f123f59:
[  0.00%] ·· Building for py3.4-humanize-mongoengine0.8.7-numpy1.9-prettytable-pytz-pyyaml-six..
[  0.00%] ·· Benchmarking py3.4-humanize-mongoengine0.8.7-numpy1.9-prettytable-pytz-pyyaml-six
[100.00%] ··· Running benchmarks.TimeSuite.time_insert_1000_events                    1.42s
```

Everyone agrees that 1.42s / 1000 events is too long. @tacaswell is working on bulk insertion as I type.

### CODE REVIEW

**Comment by ericdill** [comment]
How are you installing `nslsii_dev_configuration`?

**Comment by ericdill** [comment]
```
08:26 $ asv run
· Fetching recent changes
· Creating environments......
· Discovering benchmarks
·· Uninstalling from py3.4-humanize-mongoengine0.8.7-nslsii_dev_configuration-numpy1.9-prettytable-pytz-pyyaml-six
·· Building for py3.4-humanize-mongoengine0.8.7-nslsii_dev_configuration-numpy1.9-prettytable-pytz-pyyaml-six
·· Installing into py3.4-humanize-mongoengine0.8.7-nslsii_dev_configuration-numpy1.9-prettytable-pytz-pyyaml-six..
·· Error running /Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/bin/python /Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py discover benchmarks /var/folders/x3/507ytpw963q_8cr2343yz6200000gn/T/tmpjsh1Py
             STDOUT -------->

             STDERR -------->
             Traceback (most recent call last):
               File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py", line 658, in <module>
                 list_benchmarks(benchmark_dir, fp)
               File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py", line 639, in list_benchmarks
                 for benchmark in disc_benchmarks(root):
               File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py", line 624, in disc_benchmarks
                 for module in disc_files(root, os.path.basename(root) + '.'):
               File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py", line 613, in disc_files
                 module = import_module(package + filename)
               File "/Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/lib/python3.4/importlib/__init__.py", line 109, in import_module
                 return _bootstrap._gcd_import(name[level:], package, level)
               File "<frozen importlib._bootstrap>", line 2254, in _gcd_import
               File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
               File "<frozen importlib._bootstrap>", line 2226, in _find_and_load_unlocked
               File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
               File "<frozen importlib._bootstrap>", line 1129, in _exec
               File "<frozen importlib._bootstrap>", line 1471, in exec_module
               File "<frozen importlib._bootstrap>", line 321, in _call_with_frames_removed
               File "/Users/edill/dev/python/metadatastore/bench/benchmarks/benchmarks.py", line 4, in <module>
                 import metadatastore.commands as mds
               File "/Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/lib/python3.4/site-packages/metadatastore/__init__.py", line 1, in <module>
                 from . import conf, commands, odm_templates
               File "/Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/lib/python3.4/site-packages/metadatastore/conf.py", line 73, in <module>
                 'metadatastore', 'MDS', ['host', 'database', 'port', 'timezone'])
               File "/Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/lib/python3.4/site-packages/metadatastore/conf.py", line 69, in load_configuration
                 "file or environmental variable.".format(missing))
             KeyError: "The configuration field(s) ['port', 'database', 'timezone', 'host'] were not found in any file or environmental variable."
Traceback (most recent call last):
  File "/Users/edill/mc/envs/py2/bin/asv", line 9, in <module>
    load_entry_point('asv==0.1.1', 'console_scripts', 'asv')()
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/main.py", line 36, in main
    result = args.func(args)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/commands/__init__.py", line 47, in run_from_args
    return cls.run_from_conf_args(conf, args)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/commands/run.py", line 122, in run_from_conf_args
    skip_existing_commits=args.skip_existing_commits
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/commands/run.py", line 197, in run
    benchmarks = Benchmarks(conf, regex=bench)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmarks.py", line 285, in __init__
    for benchmark in benchmarks:
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmarks.py", line 325, in disc_benchmarks
    dots=False)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/plugins/conda.py", line 137, in run
    return self._run_executable('python', args, **kwargs)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/plugins/conda.py", line 124, in _run_executable
    os.path.join(self._path, 'bin', executable)] + args, **kwargs)
  File "/Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/util.py", line 431, in check_output
    raise ProcessError(args, retcode, stdout, stderr)
asv.util.ProcessError: Command '/Users/edill/dev/python/metadatastore/bench/env/205f559177e890f8f6cf35e974ebddd8/bin/python /Users/edill/mc/envs/py2/lib/python2.7/site-packages/asv/benchmark.py discover benchmarks /var/folders/x3/507ytpw963q_8cr2343yz6200000gn/T/tmpjsh1Py' returned non-zero exit status 1
```

**Comment by danielballan** [comment]
I have local config files for MDS and FS to fall back on.

**Comment by ericdill** [comment]
Ah, cool. thanks

File: bench/benchmarks/benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class TimeSuite:
+    """
+    An example benchmark that times the performance of various kinds
+    of iterating over dictionaries in Python.
+    """
+    def setup(self):
+        mds_setup()
+        blc_id = mds.insert_beamline_config(time=0., config_params={})
```
Comment (ericdill) [inline bench/benchmarks/benchmarks.py]: this no longer exists

File: bench/benchmarks/benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        self.rs_id = mds.insert_run_start(time=0., scan_id=1, beamline_id='bench',
+                                          beamline_config=blc_id, uid=new_uid())
```
Comment (ericdill) [inline bench/benchmarks/benchmarks.py]: drop `beamline_config=blc_id`

**Comment by danielballan** [comment]
To be continued in NSLS-II/databroker

**Pull request closed without merging**